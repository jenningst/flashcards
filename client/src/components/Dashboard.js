import React from 'react';
import styled from 'styled-components';
import LinkButton from './Elements/LinkButton';
import { Title1 } from './Elements/Text';
import { ReactComponent as MenuIcon } from '../components/icons/svg/menu.svg';
import PackCard from './PackCard';

import { GET_PACKS } from '../queries/';
import { Query } from 'react-apollo';

const UserPacks = () => (
  <Query query={GET_PACKS}>
    {({ loading, error, data }) => {
      if (loading) return 'Loading...';
      if (error) return `Error! ${error.message}`;

      const packs = data.fetchPacks;
      return (
        <React.Fragment>
          {packs.map(pack => {
            return <PackCard key={pack._id} {...pack} />
          })}
        </React.Fragment>
      );
    }}
  </Query>
);

function Dashboard() {
  return (
    <DashboardWrapper className="Dashboard">
      <Header className="Dashboard__header">
        <IconButton
          className="Dashboard__button-menu"
          data-testid="button-menu"
          onClick={() => alert('you clicked the menu!')}
        />
      </Header>
      <PackList className="Dashboard__list">
        <CreatePackButton
          className="Dashboard__button-create-pack"
          data-testid="create-pack-button"
          type="button"
          to={"/create-pack"}
        >
          <Title1>+</Title1>
        </CreatePackButton>
        
        <UserPacks />

      </PackList>
    </DashboardWrapper>
  );
};

const DashboardWrapper = styled.div`
  box-sizing: border-box;
  height: 100vh;
  overflow-y: scroll;
  padding: 1rem;
  background: ${props => props.theme.background.primary};
`;

const Header = styled.header`
  display: flex;
  justify-content: flex-end;
  color: ${props => props.theme.font.primary};
  padding: .50rem;

`;

const IconButton = styled(MenuIcon)`
  height: 2rem;
  width: 2rem;
`;

const PackList = styled.section`
  display: grid;
  overflow: scroll;
  grid-template-columns: repeat(auto-fit, minmax(135px, 45%));
  grid-gap: 1rem;
`;

// TODO: finish hover and active states
const CreatePackButton = styled(LinkButton)`
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 1rem;

  background: ${props => props.theme.background.primary};
  border: 2px solid ${props => props.theme.font.primary};
  outline: none;
  &:hover {

  }
  &:active {

  }
`;

export default Dashboard;