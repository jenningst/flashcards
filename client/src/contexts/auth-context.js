import React, { createContext, useContext } from 'react';
import Firebase from '../firebase/firebase-client';

const AuthContext = createContext(null);

function AuthProvider(props) {
  const firebase = new Firebase(); // create instance of firebase
  const data = { user: null, }; // initialize our data (for user)
  // provide firebase methods
  const registerWithEmail = (email, password) => firebase.doCreateUserWithEmailAndPassword(email, password);
  const login = (email, password) => firebase.doSignInWithEmailAndPassword(email, password);
  const signOut = () => firebase.doSignOut();
  const resetPasswordWithEmail = (email) => firebase.doResetPassword(email);
  const updatePasswordWithEmail = (password) => firebase.doUpdatePassword(password);

  return (
    <AuthContext.Provider value={
      { data,
        registerWithEmail,
        login,
        signOut,
        resetPasswordWithEmail,
        updatePasswordWithEmail,
      }
    } {...props} />
  );
};

// provides auth context to child components
/* ** IMPLEMENTED IN UserProvider ** */
function useAuth() {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within a AuthContext');
  }
  return context;
};

export { AuthProvider, useAuth };