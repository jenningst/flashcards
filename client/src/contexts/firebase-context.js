import React, { createContext, useContext } from 'react';
import Firebase from '../firebase/firebase';

const FirebaseContext = createContext(null);

// setup our pack provider
function FirebaseProvider({ children }) {
  const firebase = new Firebase();
  return (
    <FirebaseContext.Provider firebase={firebase}>
      {children}
    </FirebaseContext.Provider>
  );
};

// setup our hook to use firebase
function useFirebase() {
  const context = useContext(FirebaseContext);
  if (context === undefined) {
    throw new Error('useFirebase must be used within a FirebaseProvider');
  }
  return context;
};

export { FirebaseProvider, useFirebase };