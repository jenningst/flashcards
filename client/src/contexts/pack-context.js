import React, { createContext, useContext, useReducer } from 'react';
import { reducer, INITIAL_STATE } from '../reducers/pack-reducer';

const PackStateContext = createContext(null);
const PackDispatchContext = createContext(null);

// setup our pack provider
function PackProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  return (
    <PackStateContext.Provider value={state}>
      <PackDispatchContext.Provider value={dispatch}>
        {children}
      </PackDispatchContext.Provider>
    </PackStateContext.Provider>
  );
};

// setup our hook to use pack state
function usePackState() {
  const context = useContext(PackStateContext);
  if (context === undefined) {
    throw new Error('usePackState must be used within a PackProvider')
  }
  return context;
};

// setup our hook to use state dispatch
function usePackDispatch() {
  const context = useContext(PackDispatchContext);
  if (context === undefined) {
    throw new Error('usePackDispatch must be used within a PackProvider')
  }
  return context;
};

export { PackProvider, usePackState, usePackDispatch };