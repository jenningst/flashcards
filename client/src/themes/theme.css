/* 

button: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      active: '#5860A5',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    border: {
      primary: '#42487C',
      active: '#939CE8'
      disabled: '#b1aeb7',
      input: '#e9eaf0',
    }, // TODO: consolidate everything below; using the above
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }

}*/