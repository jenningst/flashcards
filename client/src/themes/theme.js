export const darkTheme = {
  font: {
    primary: '#FFFFFF',
    secondary: '#E0E0E0',
    tertiary: '#666666',
  },
  background: {
    primary: '#282626',
    secondary: '#313131',
    highlight: '#6121BF',
    special: '#7358F7',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};

export const lightTheme = {
  color: {
    button: {
      primary: '#939CE8',
      primaryHover: '#7883E2',
      active: '#5860A5',
      secondary: '#50cdff',
      secondaryHover: '#6FD6FF',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#33364d',
      blackCoral: '#585A6D',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1',
      placeholder: '#e9eaf0',
      danger: '#DE6A60',
    },
    border: {
      primary: '#5860A5',
      secondary: '#7883E2',
      disabled: '#b1aeb7',
      inputDefault: '#e9eaf0',
      transparent: '#0000ffff',
      eerieBlack: '#16171D',
      charleston: '#33364d',
    }, // TODO: consolidate everything below; using the above
    background: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
    },
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }
  },
  spacing: {
    compact: '1rem',
    comfortable: '1.5rem',
    generous: '2rem',
  },
  font: {
    primary: '#4e5167',
    secondary: '#262626',
    tertiary: '#666666',
  },
  background: {
    primary: '#ffffff',
    secondary: '#fcfbfc',
    highlight: '#9388ff',

    special: '#50cdff',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};