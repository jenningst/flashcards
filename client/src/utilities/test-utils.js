const uuidv4 = require('uuid/v4');

function seedMockPack(pack, n = 1) {
  const filter = uuidv4(); // generate a filter
  const name = uuidv4(); // generate a name
  const modes = ['REVIEW_MODE', 'WRITE_MODE']; // TODO: remove hard-coding
  const mode = modes[getRandomInt(0, 1)]; // generate a random mode
  let numToSeed = getRandomInt(1, n || pack.length - 1); // generate a random number of cards
  
  let cards = [];
  for (let i = 0; i < numToSeed; i++) {
    let card = getRandomElement(pack, pack.length - 1);
    card.id = uuidv4();
    cards.push(card);
  }
 
  return { mode, name, filter, cards };
};

function getRandomElement(arr, ceiling) {
  return arr[getRandomInt(0, ceiling)];
};

function getRandomInt(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min;
};

export { seedMockPack };