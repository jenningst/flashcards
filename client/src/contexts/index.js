import React from 'react';
import { AuthProvider } from './auth-context';
import { UserProvider } from './user-context';
import { ApolloProvider } from 'react-apollo';

import { ApolloClient, HttpLink } from 'apollo-boost';
import { InMemoryCache } from 'apollo-cache-inmemory';

const client = new ApolloClient({
  link: new HttpLink({ uri: "http://localhost:4000/graphql" }),
  cache: new InMemoryCache(),
});

function AppProviders({ children }) {
  return (
    <AuthProvider>
      <UserProvider>
        <ApolloProvider client={client}>{children}</ApolloProvider>
      </UserProvider>
    </AuthProvider>
  );
};

export default AppProviders;