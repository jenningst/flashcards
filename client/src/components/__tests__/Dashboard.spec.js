import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { MockedProvider } from 'react-apollo/test-utils';
import { ThemeProvider } from 'styled-components';
import { darkTheme, lightTheme } from '../../themes/theme';

import {
  cleanup,
  render,
} from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import Dashboard from '../Dashboard';
import { AuthProvider } from '../../contexts/auth-context';

const mockAuth = {
  user: {
    uid: '951affbb-af5e-4583-bc4d-e8e419a1e1cc',
    email: 'test@test.com',
  },
  initialising: false,
  error: null,
};

beforeEach(() => { });
afterEach(cleanup);

const renderComponent = ({ theme, auth }) =>
  act(() => {
    render(
      <AuthProvider value={auth}>
        <ThemeProvider theme={theme}>
          <MockedProvider>
            <MemoryRouter>
              <Dashboard />
            </MemoryRouter>
          </MockedProvider>
        </ThemeProvider>
      </AuthProvider>
    );
  })

describe('<Dashboard /> spec', () => {
  it('assert component matches snapshot', () => {
    const { container } = renderComponent({ 
      theme: lightTheme,
      auth: mockAuth,
    });
    expect(container.firstChild).toMatchSnapshot();
  });
  
  it('assert the add collection button renders', () => {
    const { getByTestId } = renderComponent({ theme: lightTheme });
    expect(getByTestId('create-pack-button')).toBeEnabled();
  });

  it.todo('assert a menu icon renders');
  it.todo('assert menu icon handler is called');
  it.todo('assert user avatar renders');
  it.todo('assert user avatar handled is called');

  it.todo('assert can receive packs from gql query');
  it.todo('assert renders PackCards from gql query');
});