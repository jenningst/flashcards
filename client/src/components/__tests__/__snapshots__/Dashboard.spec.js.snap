// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> spec assert component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Dashboard sc-gZMcBi eDycXU"
  >
    <header
      class="Dashboard__header sc-gqjmRU fTNAVy"
    >
      <svg
        class="Dashboard__button-menu sc-VigVT fCGeOx"
        data-testid="button-menu"
      >
        menu.svg
      </svg>
    </header>
    <section
      class="Dashboard__list sc-jTzLTM FholT"
    >
      <button
        class="Dashboard__button-create-pack sc-fjdhpX kSPajq"
        data-testid="create-pack-button"
        type="button"
      >
        <h1
          class="sc-bdVaJa beESmM"
        >
          +
        </h1>
      </button>
      Loading...
    </section>
  </div>
</DocumentFragment>
`;
